core/src/main/java/org/apache/calcite/sql/SqlNumericLiteral.java

```diff
@@ -99,15 +99,22 @@ public boolean isExact() {
     if (isExact) {
       int scaleValue = requireNonNull(scale, "scale");
       if (0 == scaleValue) {
-        BigDecimal bd = getValueNonNull();
-        SqlTypeName result;
-        long l = bd.longValue();
-        if ((l >= Integer.MIN_VALUE) && (l <= Integer.MAX_VALUE)) {
-          result = SqlTypeName.INTEGER;
-        } else {
-          result = SqlTypeName.BIGINT;
+        try {
+          BigDecimal bd = getValueNonNull();
+          SqlTypeName result;
+          // Will throw if the number cannot
+          // be represented as a long.
+          long l = bd.longValue();
+          if ((l >= Integer.MIN_VALUE) && (l <= Integer.MAX_VALUE)) {
+            result = SqlTypeName.INTEGER;
+          } else {
+            result = SqlTypeName.BIGINT;
+          }
+          return typeFactory.createSqlType(result);
+        } catch (ArithmeticException ex) {
+          // This indicates that the value does
+          // not fit in any integer type.  Fallback to DECIMAL.
```

Comment:
Same as above we don't need break line and before `Fallback`  there are extra spaces.